### DISCUSSION TITLE
Convert a tess file into an stl or iges file

### DISCUSSION THREAD

**Discussion by ap264** [neperfepx/neper#537] [Q&A] [answered]
Hi

I have a 3d tessellation file which consists of 79 grains with unique equivalent diameter. The tessellation in enveloped in a cube of length 1.3. I would like to know if it is possible for me to convert this file into a stl or an iges file so that i can export it into abaqus and perform further analysis. I came across some GITHUB codes to resolve this complexity but they are catered towards parellelpied and columnar tessellations. 

Thank you    
![img5](https://user-images.githubusercontent.com/109927295/185011085-37e96c16-48c1-44f0-83f3-11262c26ac3d.png)

**Comment by rquey** [comment, marked as answer]
Neper does output some STL by itself, using `-format stl`.

**Reply by ap264** [reply]
Dear Sir

First of all, I would like to thank you for creating this software and responding to my request. 

I used the STL format and exported it to Solid works. But the resulting geometry is just one square and Voronoi polygons are just observed to be mesh elements (image attached). I would like to know if there is a way to export the Voronoi polygon in an STL format while displaying the vornoi cells as it should be. 
![solidworks](https://user-images.githubusercontent.com/109927295/186496730-877989ad-3592-444f-8fe5-0ddc91b222fb.PNG)

Also, I have another question, I have modelled the tessellation file and meshed it using the following commands
neper -T n -100 -reg 1 -o mytess
neper -M mytess.tess -dim 2 -elttype tri  -rcl 0.6 -meshface "domtype!=2" -fasets none -nsets none -format inp -o mymsh 

the mesh is then exported to Abaqus (image attached)
![abaqus](https://user-images.githubusercontent.com/109927295/186497654-3fdaf806-3180-4e23-a1e6-ac3e63aee6fc.PNG)

But i get this error in Abaqus when I proceed to check the mesh quality (image attached)
![abaqus error](https://user-images.githubusercontent.com/109927295/186497953-dc2a6a80-9acf-4e9b-a674-f8d5e29e1cf0.PNG)

I would like to know if there is a way to improve the mesh within the neper software to rectify this error

**Reply by rquey** [reply]
> Dear Sir
> 
> First of all, I would like to thank you for creating this software and responding to my request.
> 
> I used the STL format and exported it to Solid works. But the resulting geometry is just one square and Voronoi polygons are just observed to be mesh elements (image attached). I would like to know if there is a way to export the Voronoi polygon in an STL format while displaying the vornoi cells as it should be. ![solidworks](https://user-images.githubusercontent.com/109927295/186496730-877989ad-3592-444f-8fe5-0ddc91b222fb.PNG)
> 

Neper divides polygonal faces into triangles. I don't know if there is a way to write polygons in STL format.

> Also, I have another question, I have modelled the tessellation file and meshed it using the following commands neper -T n -100 -reg 1 -o mytess neper -M mytess.tess -dim 2 -elttype tri -rcl 0.6 -meshface "domtype!=2" -fasets none -nsets none -format inp -o mymsh
> 
> the mesh is then exported to Abaqus (image attached) ![abaqus](https://user-images.githubusercontent.com/109927295/186497654-3fdaf806-3180-4e23-a1e6-ac3e63aee6fc.PNG)
> 
> But i get this error in Abaqus when I proceed to check the mesh quality (image attached) ![abaqus error](https://user-images.githubusercontent.com/109927295/186497953-dc2a6a80-9acf-4e9b-a674-f8d5e29e1cf0.PNG)
> 
> I would like to know if there is a way to improve the mesh within the neper software to rectify this error

This is probably the best you can get for a (regularized) Voronoi tessellation (which is what your -T command generates). Aren't the Abaqus messages more warnings than errors?
Better mesh quality is obtained for other morphologies, especially when the cell sphericities are higher (as in `gg`).

**Reply by ap264** [reply]
from your reply: Aren't the Abaqus messages more warnings than errors? 

The Abaqus gives me this message when I am evaluating the mesh quality. But when I submit the job for analysis, Abaqus stop the simulation in its early stages and gives me an output error message saying that some mesh elements have negative or small areas. 

from your reply: Better mesh quality is obtained for other morphologies, especially when the cell sphericities are higher (as in gg).

Sure sir. I will check this out. I have a question regarding sphericity. According to you, what is the optimum sphericity to achieve a better mesh quality? Also, are there other morphology options in the software which can improve my mesh quality?

**Reply by ap264** [reply]
Dear Sir

Just FYI. I used the morphology with the sphericity option. I meshed my Voronoi polygon with tetrahedral elements and it is working when ii export it to Abaqus. But when I switch to 2d mesh with the -dim 2 option instead tetrahedron then my mesh in Abaqus is giving me the same error.  

Also, I would like to know if I could mesh my tessellation model with continuum shell elements like S3R elements in Neper?

**Reply by rquey** [reply]
With regards to Abaqus, I'm afraid I can't help.

In general, the higher the grain sphericities, the better the mesh quality. `-morpho centroidal` may be a good way to get spherical grains, if you don't really matter about the morphology...

**Reply by ap264** [reply]
Understood sir. Thank you for the input.
